src/System.Linq.Expressions/src/System/Linq/Expressions/StackGuard.cs

```diff
@@ -0,0 +1,91 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Linq.Expressions
+{
+    internal sealed class StackGuard
+    {
+        private const int StackProbingThreshold = 100;
+        private const int MaxExecutionStackCount = 1024;
+
+        private int _recursionDepth;
+        private int _executionStackCount;
+
+        public bool TryEnterOnCurrentStack()
+        {
+            if (++_recursionDepth >= StackProbingThreshold)
+            {
+                try
+                {
+                    RuntimeHelpers.EnsureSufficientExecutionStack();
+                }
+                catch (InsufficientExecutionStackException) when (_executionStackCount < MaxExecutionStackCount)
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        public void Exit()
+        {
+            _recursionDepth--;
+        }
+
+        public void RunOnEmptyStack(Action<object> action, object state)
+        {
```

Comment:
Good suggestion; it does add four overloads though (2-3 arguments and void/non-void returning), so I'm offsetting the increased code size by using a single core implementation using `StartNew<R>(Func<object, R>, object, ...)` and making the `Action<>` based overloads return `default(object)` from the body. I could really return any object from the function; is there any preferred type to close a `Task<R>` over when it merely represents a sentinel value?

```csharp
public void RunOnEmptyStack<T1, T2>(Action<T1, T2> action, T1 arg1, T2 arg2)
{
    RunOnEmptyStackCore(s =>
    {
        var t = (Tuple<Action<T1, T2>, T1, T2>)s;
        t.Item1(t.Item2, t.Item3);
        return default(object);
    }, Tuple.Create(action, arg1, arg2));
}
```